Fix downgrades from 10.2-4 to 10.2-2

hanefi opened the pull request:
DESCRIPTION: Fixes a bug in `ALTER EXTENSION citus UPDATE`

We had a series of changes on columnar that made it impossible for a Citus user to downgrade from 10.2-4 to 10.2-2. Since we test downgrades to immediate previous versions, we did not capture this in our tests. Here are the series of changes.

- `10.2-1` introduced a btree index named `columnar.stripe_first_row_number_idx`
- `10.2-3` had a unique index with the same name. To accomplish that, we dropped the btree index, and create a unique index with the same name.
- `10.2-4` introduced `columnar_ensure_am_depends_catalog()` that adds pg_depend entries so that Columnar access method depended on objects such as `stripe_first_row_number_idx`

If a user upgrades to `>=10.2-4` we create a dependency record, and this prevents users from downgrading to an earlier version than `10.2-3` since the downgrade file `columnar--10.2-3--10.2-2.sql` wanted to drop the unique index and create a btree index instead. However this created an error because columnar am depended on that index.

We do not usually like to update earlier migration versions, but there is no other solution that I could think of. 

## Notes to reviewer:

Consider reviewing the commits one by one.
- Commit#1 aims to improve downgrade scripts overall.
- Commit#2 documents the failure
- Commit#3 fixes the problem by updating all the files that attempted to drop `stripe_first_row_number_idx` index.

Related: #6041

onurctirtir reviewed (commented):
Looks good to me. Even though we're modifying some of the existing downgrade scripts, I'm okay with this approach since backporting those changes would be hard a lot.

onurctirtir reviewed (approved):
.. So as we discussed, let's ask for an approval from @thanodnl or @onderkalaci too.

thanodnl commented on src/backend/columnar/sql/columnar--10.2-2--10.2-3.sql:
```diff
@@ -10,6 +10,17 @@
+WHERE classid = 'pg_am'::regclass::oid
+    AND objid IN (select oid from pg_am where amname = 'columnar')
+    AND objsubid = 0
+    AND refclassid = 'pg_class'::regclass::oid
```
might be worth to fully classify this entry as well, guess it's `pg_catalog`, to prevent overloads, although it might be that you can't really create `pg_` prefixed tables anyway. Still safe to qualify it here.

hanefi replied:
The function that inserted those rows did not qualify this part. I agree that it makes sense to qualify this. However, I think it is more important to have a consistency between the two scripts.

I think we should either update both [columnar_ensure_am_depends_catalog](https://github.com/citusdata/citus/blob/580ab012bf04710604bef74a774a9f8afb656130/src/backend/columnar/sql/udfs/columnar_ensure_am_depends_catalog/latest.sql) and here or none.

thanodnl commented on src/backend/columnar/sql/columnar--10.2-2--10.2-3.sql:
```diff
@@ -10,6 +10,17 @@
 -- To do that, drop stripe_first_row_number_idx and create a unique
 -- constraint with the same name to keep the code change at minimum.
+--
+-- If we have a pg_depend entry for this index, we can not drop it as
+-- the extension depends on it. Remove the pg_depend entry if it exists.
```
I am a bit baffled that the extension depends on an index instead of the index having an extension dependency to the extension preventing it from being dropped outside of `DROP EXTENSION`.

Could you explain it a bit better why a columnar catalog table has an index depending on the extension?

thanodnl commented on src/backend/columnar/sql/columnar--10.2-2--10.2-3.sql:
```diff
@@ -10,6 +10,17 @@
 --
```
What Citus versions has this been released in? Do we need to backport this? What would happen if any of the previously released Citus versions is still used somewhere. Guess that it will fail directly during the `ALTER EXTENSION citus UPDATE;` ? Which probably would suffice to have a patch release for 10.2, 11.0 and 11.1?

thanodnl commented on src/test/regress/sql/multi_extension.sql:
```diff
@@ -400,13 +400,21 @@ SELECT * FROM multi_extension.print_extension_changes();
 
--- Snapshot of state at 10.2-5
-ALTER EXTENSION citus UPDATE TO '10.2-5';
+-- There was a bug when downgrading to 10.2-2 from 10.2-4
+-- Test that we do not have any issues with this particular downgrade
+ALTER EXTENSION citus UPDATE TO '10.2-2';
+ALTER EXTENSION citus UPDATE TO '10.2-4';
```
Above we test a downgrade from `10.2-4` to `10.2-3` and a separate one for `10.2-3` to `10.2-2`. How is a direct downgrade from `10.2-4` to `10.2-2` causing an issue that we didn't run into above?

and actually there is a test:
```
-- Make sure that we don't delete pg_depend entries added in
-- columnar--10.2-3--10.2-4.sql when downgrading to 10.2-3.
SELECT COUNT(*)=10
FROM pg_depend
WHERE classid = 'pg_am'::regclass::oid AND
      objid = (select oid from pg_am where amname = 'columnar') AND
      objsubid = 0 AND
      refclassid = 'pg_class'::regclass::oid AND
      refobjsubid = 0 AND
      deptype = 'n';
```

Which seems to be specifically testing the invariant you are changing here.

To say the least, I am confused.

thanodnl reviewed (approved):
Thanks for the clarification, it makes a bit more sense to me now.

I guess we deliberately wanted the pg_depend entries to stay after a downgrade to the previous version, as it is the right thing to have when performing a pg_upgrade.

Except that we didn't anticipate that a version earlier we actually drop the dependant object on a downgrade, meaning you can't downgrade further anymore.

Given how unlikely downgrades are, and that we actually don't include downgrade scripts in our packages I wonder if we do need to backport these changes at all. Although there seems to be 1 fix in 10.2-2 to 10.2-3, which if correct needs to be backported to the minor versions that had this file included, since it is in the upgrade path for users coming from before 10.2-2.

Besides this whole dance, I wonder if our version checking is enough now that we more often needing to ship sql changes in patch versions. I don't think we complain for any patch version of 10.2.x if it introduced new `10.2-y` sql versions. I believe this only checks the major + minor version, which got broken the first time we released sql changes in a patch.

After merging this, do you have time to look into this, and see if it would make sense to increase the version check to the declared sql version in the `citus.control` file we ship?

hanefi commented:
> Besides this whole dance, I wonder if our version checking is enough now that we more often needing to ship sql changes in patch versions. I don't think we complain for any patch version of 10.2.x if it introduced new `10.2-y` sql versions. I believe this only checks the major + minor version, which got broken the first time we released sql changes in a patch.

I do not think that I follow what gets broken here. We need to make a deliberate effort to make sure we do not break stuff when we need to introduce a new patch version to an older major version. We do not have a test architecture that covers this and I try to own either authoring or reviewing&testing complex backports with migration files.

For example in #6100 we had a bugfix that needs a schema change in a migration file. We also wanted to backport this to earlier versions. This required us to squeeze in several new migration versions in 10.2 as well as 11.0.

Backporting efforts for #6100 is in #6176. When backporting, we introduced `10.2-5` in after we already shipped several 11.0 packages. To be able to do that, we did the following:
- created `citus--10.2-4--10.2-5.sql` with the relevant fix
- renamed `citus--10.2-4--11.0-1.sql` to `citus--10.2-5--11.0-1.sql`
- made sure that all the migration files are on `release-10.2` and all later release branches as well as main branch.
- made sure that `citus.control` is bumped to `10.2-5` on the `release-10.2` branch.

This set of changes was ok. However these complex changes are not really covered in our tests. Today, we only test upgrades and downgrades in the currently checked out branch.

> After merging this, do you have time to look into this, and see if it would make sense to increase the version check to the declared sql version in the `citus.control` file we ship?

Can you explain what you mean by "increasing  version check` ? Do you want me to add tests for downgrading from the version specified in `citus.control` (10.2-5) down to 10.2-2 ?
